documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md 

```diff
@@ -0,0 +1,82 @@
+- As a provider of openly-licensed media, I want to know what media properties Openverse collects so that I can change my API responses.
+
+The information on data types, transformations, and validations will be presented as a table and will be extracted from the code.  Other information about the data selection properties and data discrepancies will be written in a separate markdown file as it is easier to read, write, and lint.
+
+The catalog has a DAG documentation generator that extracts the docstrings from the DAGs and creates a page with the documentation. We can use the same approach to create a documentation page for the media properties. The DAG doc generator runs the code inside docker to extract DAG information. In this project, we can simplify and use the `ast` module to parse static code and extract the docstrings to make the checks faster. This will, however, mean that some of the settings will not be picked up and will have to be updated during the generation. One example of this is the `nullable` field in the `Column` class: it falls back to the value of `not nullable` during the object initiation.
+
+## Outlined Steps
+
+### `just` script
+
+The script parses the SQL DDL files, Python files and markdown files to generate a markdown page with a table and long-form notes. This script can `diff` with the existing documentation page (created in step 4 below) for the `precommit` check and post a note about how to update the docs if differences are detected.
+
+### 1. SQL DDL parsing
+We will use the names of the properties as they are in the upstream database, in the same order as they are in the `image_view`/`audio_view` materialized views. These views are selected over the main `image` and `audio` tables because they are used by the API during data refresh and contain a complete list of properties.
+
+A Python script is written to extract the information from the SQL files into a mapping with the name of the media property and the following values: SQL datatype, and SQL constraints. More data will be added to this mapping in the further steps.
+
+### 2. Parsing of the Python code
+The `Column` class and its child classes describe the validations we use to write the data collected by provider scripts. The `add_item` method has docstrings with short descriptions of what each property is. The information from these items will be added into the mapping from step 1.
+
+### 3. Adding the long-form information from the markdown file
+The markdown file will be parsed and split into the preamble, postamble (???) and the information about each data property. This information would be added to the `description` in the property mapping.
+
+### 4. Writing the markdown file
+The mapping will be used to create a new markdown file: the Preamble, the tables for Audio and Image, the long-form document from step 3.
```

Comment:
I think I'm confused about the files written here. In the overview you wrote:


> The information on data types, transformations, and validations will be presented as a table and will be extracted from the code.  Other information about the data selection properties and data discrepancies will be written in a separate markdown file as it is easier to read, write, and lint.

Is step 3 referring to generating a long form file with 'other information' (the second thing listed above), and step 4 is referring to a separate document with just the tables?

> The mapping will be used to create a new markdown file: the Preamble, the tables for Audio and Image, the long-form document from step 3.

Does that last part mean the the long-form document is contained within the new markdown file? Or linked?

Reply:
Step 4 is referring to the actual documentation page that would be posted to docs.openverse.org This will be auto-generated, and will use the data from steps 1-3 to create something like this:

# Openverse Media Properties

preamble (copied from `general.md`)

## Overview of the media properties

table with the datatype details, possibly with links to the headings below. Autogenerated in steps 1-2.

## In-depth description

### identifier
Description of the `identifier` field. Step 3 reads the manually-written `media_properties.md` file, splits it by the media property, and then writes it here in the correct order.

### title 
description

# Other considerations (postamble)
## Encoding problems
Description of what happened with encoding of tags and titles

Reply:
Got it, this comment and [this one](https://github.com/WordPress/openverse/pull/1904/files#r1179918049) totally cleared up my confusion. Thanks!